src/sampler/record.rs

```diff
@@ -0,0 +1,451 @@
+                    config.sample_rate,
+                    timing_error_traces.load(Ordering::Relaxed),
+                    total_traces.load(Ordering::Relaxed),
+                )?;
+                summary_time = std::time::Instant::now() + Duration::from_secs(1);
+            }
+        }
+    }
+
+    // Finish writing all data to disk
+    if config.out_path.display().to_string() == "-" {
+        out.complete(&mut std::io::stdout())?;
+    } else {
+        let mut out_file = File::create(&config.out_path).context(format!(
+            "Failed to create output file {}",
+            &config.out_path.display()
+        ))?;
+        out.complete(&mut out_file)?;
+    }
+    raw_store.complete();
+
+    // Check for errors from the child threads. Ignore errors unless every single thread
+    // returned an error. If that happens, return the last error. This lets rbspy successfully
+    // record processes even if the parent thread isn't a Ruby process.
+    let mut num_ok = 0;
+    let mut last_result = Ok(());
+    for result in result_receiver.iter() {
+        if result.is_ok() {
+            num_ok += 1;
+        }
+        last_result = result;
+    }
+
+    match num_ok {
+        0 => last_result,
+        _ => Ok(()),
+    }
+}
+
+/// Records stack traces and sends them to a channel in another thread where they can be aggregated
+fn record(
+    pid: Pid,
+    sample_rate: u32,
+    maybe_stop_time: Option<Instant>,
+    done: Arc<AtomicBool>,
+    timing_error_traces: Arc<AtomicUsize>,
+    total_traces: Arc<AtomicUsize>,
+    sender: SyncSender<StackTrace>,
+    lock_process: bool,
+) -> Result<(), Error> {
+    let mut getter = initialize(pid, lock_process)?;
+
+    let mut total = 0;
+    let mut errors = 0;
+
+    let mut sample_time = SampleTime::new(sample_rate);
+    #[cfg(windows)]
+    {
+        // This changes a system-wide setting on Windows so that the OS wakes up every 1ms
+        // instead of the default 15.6ms. This is required to have a sleep call
+        // take less than 15ms, which we need since we usually profile at more than 64hz.
+        // The downside is that this will increase power usage: good discussions are:
+        // https://randomascii.wordpress.com/2013/07/08/windows-timer-resolution-megawatts-wasted/
+        // and http://www.belshe.com/2010/06/04/chrome-cranking-up-the-clock/
+        unsafe {
+            timeapi::timeBeginPeriod(1);
+        }
+    }
+
+    while !done.load(Ordering::Relaxed) {
+        total += 1;
+        let trace = getter.get_trace();
+        match trace {
+            Ok(ok_trace) => {
+                sender.send(ok_trace)?;
+            }
+            Err(x) => {
+                if let Some(MemoryCopyError::ProcessEnded) = x.downcast_ref() {
+                    debug!("Process {} ended", pid);
+                    return Ok(());
+                }
+
+                errors += 1;
+                if errors > 20 && (errors as f64) / (total as f64) > 0.5 {
```

Comment:
What if there will be less than 20 errors? I guess those would not be printed. Maybe there should be a check if any error occurred, just after the loop and then they can be printed?

Reply:
Yeah, we don't have a nice way to report non-fatal errors right now. Would be nice. I'll keep this in mind too.